File: apps/demo-be-python-fastapi/tests/unit/steps/token_lifecycle_steps.py
"""Unit BDD step definitions for token lifecycle feature."""

import pytest
from fastapi.testclient import TestClient
from pytest_bdd import given, scenarios, then, when

from demo_be_python_fastapi.auth.jwt_service import create_expired_refresh_token
from tests.unit.conftest import GHERKIN_ROOT

pytestmark = pytest.mark.unit

scenarios(str(GHERKIN_ROOT / "authentication" / "token-lifecycle.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token and refresh token',
    target_fixture="alice_tokens",
)
def alice_login_tokens(client: TestClient, registered_user: dict) -> dict:
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": "alice", "password": _PASSWORD},
    )
    assert resp.status_code == 200, f"Login failed: {resp.text}"
    return resp.json()


@given("alice's refresh token has expired", target_fixture="alice_tokens")
def alice_expired_refresh_token(client: TestClient, registered_user: dict) -> dict:
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": "alice", "password": _PASSWORD},
    )
    assert resp.status_code == 200
    tokens = resp.json()
    tokens["refresh_token"] = create_expired_refresh_token(registered_user["id"])
    return tokens


@given(
    "alice has used her refresh token to get a new token pair",
    target_fixture="alice_tokens",
)
def alice_used_refresh_token(client: TestClient, registered_user: dict) -> dict:
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": "alice", "password": _PASSWORD},
    )
    assert resp.status_code == 200
    original_tokens = resp.json()
    refresh_resp = client.post(
        "/api/v1/auth/refresh",
        json={"refresh_token": original_tokens["refresh_token"]},
    )
    assert refresh_resp.status_code == 200
    return {**original_tokens, "_original_refresh": original_tokens["refresh_token"]}


@given('the user "alice" has been deactivated', target_fixture="deactivated_alice")
def deactivate_alice(client: TestClient, registered_user: dict, alice_tokens: dict) -> dict:
    resp = client.post(
        "/api/v1/users/me/deactivate",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )
    assert resp.status_code == 200
    return registered_user


@given("alice has already logged out once")
def alice_already_logged_out(client: TestClient, registered_user: dict, alice_tokens: dict) -> None:
    client.post(
        "/api/v1/auth/logout",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )


# --- When steps ---


@when("alice sends POST /api/v1/auth/refresh with her refresh token", target_fixture="response")
def alice_refresh(client: TestClient, alice_tokens: dict):  # type: ignore[no-untyped-def]
    return client.post(
        "/api/v1/auth/refresh",
        json={"refresh_token": alice_tokens["refresh_token"]},
    )


_STEP_REFRESH_ORIGINAL = "alice sends POST /api/v1/auth/refresh with her original refresh token"


@when(_STEP_REFRESH_ORIGINAL, target_fixture="response")
def alice_refresh_with_original(client: TestClient, alice_tokens: dict):  # type: ignore[no-untyped-def]
    original = alice_tokens.get("_original_refresh", alice_tokens["refresh_token"])
    return client.post(
        "/api/v1/auth/refresh",
        json={"refresh_token": original},
    )


@when("alice sends POST /api/v1/auth/logout with her access token", target_fixture="response")
def alice_logout(client: TestClient, alice_tokens: dict):  # type: ignore[no-untyped-def]
    return client.post(
        "/api/v1/auth/logout",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )


@when("alice sends POST /api/v1/auth/logout-all with her access token", target_fixture="response")
def alice_logout_all(client: TestClient, alice_tokens: dict):  # type: ignore[no-untyped-def]
    return client.post(
        "/api/v1/auth/logout-all",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )


# --- Then steps ---


@then("alice's access token should be invalidated")
def check_access_token_invalidated(client: TestClient, alice_tokens: dict) -> None:
    resp = client.get(
        "/api/v1/users/me",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )
    assert resp.status_code == 401, f"Expected 401 after logout, got {resp.status_code}"

File: apps/demo-fs-ts-nextjs/test/unit/be-steps/token-lifecycle.steps.ts
import path from "path";
import { loadFeature, describeFeature } from "@amiceli/vitest-cucumber";
import { expect } from "vitest";
import { SignJWT } from "jose";
import { createTestContext, registerUser, loginUser, getAuth, type TestContext } from "./helpers/test-context";

const feature = await loadFeature(
  path.resolve(__dirname, "../../../../../specs/apps/demo/be/gherkin/authentication/token-lifecycle.feature"),
);

const JWT_SECRET = process.env.APP_JWT_SECRET ?? "test-jwt-secret-at-least-32-chars-long!!";

describeFeature(feature, ({ Scenario, Background }) => {
  let ctx: TestContext;

  Background(({ Given, And }) => {
    Given("the API is running", () => {
      ctx = createTestContext();
    });

    And('a user "alice" is registered with password "Str0ng#Pass1"', async () => {
      await registerUser(ctx, "alice", "alice@example.com", "Str0ng#Pass1");
    });

    And('"alice" has logged in and stored the access token and refresh token', async () => {
      await loginUser(ctx, "alice", "Str0ng#Pass1");
    });
  });

  Scenario("Successful refresh returns a new access token and refresh token", ({ When, Then, And }) => {
    When("alice sends POST /api/v1/auth/refresh with her refresh token", async () => {
      const refreshToken = ctx.tokens.get("alice_refresh")!;
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/refresh", { refreshToken }, null);
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });

    And('the response body should contain a non-null "accessToken" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).accessToken).toBeDefined();
    });

    And('the response body should contain a non-null "refreshToken" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).refreshToken).toBeDefined();
    });
  });

  Scenario("Reject refresh with an expired refresh token", ({ Given, When, Then, And }) => {
    Given("alice's refresh token has expired", async () => {
      const secret = new TextEncoder().encode(JWT_SECRET);
      const expiredToken = await new SignJWT({ sub: ctx.userIds.get("alice"), tokenType: "refresh" })
        .setProtectedHeader({ alg: "HS256" })
        .setIssuedAt(Math.floor(Date.now() / 1000) - 7200)
        .setExpirationTime(Math.floor(Date.now() / 1000) - 3600)
        .setJti(crypto.randomUUID())
        .setIssuer("demo-fs-ts-nextjs")
        .sign(secret);
      ctx.tokens.set("alice_refresh", expiredToken);
    });

    When("alice sends POST /api/v1/auth/refresh with her refresh token", async () => {
      const refreshToken = ctx.tokens.get("alice_refresh")!;
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/refresh", { refreshToken }, null);
    });

    Then("the response status code should be 401", () => {
      expect(ctx.response!.status).toBe(401);
    });

    And("the response body should contain an error message about token expiration", () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).length).toBeGreaterThan(0);
    });
  });

  Scenario("Original refresh token is rejected after rotation (single-use)", ({ Given, When, Then, And }) => {
    Given("alice has used her refresh token to get a new token pair", async () => {
      const refreshToken = ctx.tokens.get("alice_refresh")!;
      ctx.context.originalRefresh = refreshToken;
      const resp = await ctx.client.dispatch("POST", "/api/v1/auth/refresh", { refreshToken }, null);
      const body = resp.body as { accessToken: string; refreshToken: string };
      ctx.tokens.set("alice_access", body.accessToken);
      ctx.tokens.set("alice_refresh", body.refreshToken);
    });

    When("alice sends POST /api/v1/auth/refresh with her original refresh token", async () => {
      ctx.response = await ctx.client.dispatch(
        "POST",
        "/api/v1/auth/refresh",
        { refreshToken: ctx.context.originalRefresh as string },
        null,
      );
    });

    Then("the response status code should be 401", () => {
      expect(ctx.response!.status).toBe(401);
    });

    And("the response body should contain an error message about invalid token", () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).length).toBeGreaterThan(0);
    });
  });

  Scenario("Refresh fails for a deactivated user", ({ Given, When, Then, And }) => {
    Given('the user "alice" has been deactivated', async () => {
      const auth = getAuth(ctx, "alice");
      await ctx.client.dispatch("POST", "/api/v1/users/me/deactivate", null, auth);
    });

    When("alice sends POST /api/v1/auth/refresh with her refresh token", async () => {
      const refreshToken = ctx.tokens.get("alice_refresh")!;
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/refresh", { refreshToken }, null);
    });

    Then("the response status code should be 401", () => {
      expect(ctx.response!.status).toBe(401);
    });

    And("the response body should contain an error message about account deactivation", () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("deactivat");
    });
  });

  Scenario("Logout current session invalidates the access token", ({ When, Then, And }) => {
    When("alice sends POST /api/v1/auth/logout with her access token", async () => {
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/logout", null, getAuth(ctx, "alice"));
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });

    And("alice's access token should be invalidated", async () => {
      const resp = await ctx.client.dispatch("GET", "/api/v1/users/me", null, getAuth(ctx, "alice"));
      expect(resp.status).toBe(401);
    });
  });

  Scenario("Logout all devices invalidates tokens from all sessions", ({ When, Then, And }) => {
    When("alice sends POST /api/v1/auth/logout-all with her access token", async () => {
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/logout-all", null, getAuth(ctx, "alice"));
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });

    And("alice's access token should be invalidated", async () => {
      const resp = await ctx.client.dispatch("GET", "/api/v1/users/me", null, getAuth(ctx, "alice"));
      expect(resp.status).toBe(401);
    });
  });

  Scenario("Logout is idempotent — repeating logout on the same token returns 200", ({ Given, When, Then }) => {
    Given("alice has already logged out once", async () => {
      await ctx.client.dispatch("POST", "/api/v1/auth/logout", null, getAuth(ctx, "alice"));
    });

    When("alice sends POST /api/v1/auth/logout with her access token", async () => {
      ctx.response = await ctx.client.dispatch("POST", "/api/v1/auth/logout", null, getAuth(ctx, "alice"));
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });
  });
});
